implementations/rust/ockam/ockam_identity/src/secure_channel/handshake/handshake.rs

```diff
@@ -0,0 +1,641 @@
+    }
+
+    /// Read the message 3 encrypted key at the beginning of the message
+    fn read_message3_encrypted_key(message: &Vec<u8>) -> Result<&[u8]> {
+        Self::read_start(message, Self::encrypted_key_size())
+    }
+
+    /// Read the message 3 payload which is present after the encrypted key
+    fn read_message3_payload(message: &Vec<u8>) -> Result<&[u8]> {
+        Self::read_end(message, Self::encrypted_key_size())
+    }
+
+    /// Read the first 'length' bytes of the message
+    fn read_start(message: &Vec<u8>, length: usize) -> Result<&[u8]> {
+        if message.len() < length {
+            return Err(XXError::MessageLenMismatch.into());
+        }
+        Ok(&message[0..length])
+    }
+
+    /// Read the bytes of the message after the first 'drop_length' bytes
+    fn read_end(message: &Vec<u8>, drop_length: usize) -> Result<&[u8]> {
+        if message.len() < drop_length {
+            return Err(XXError::MessageLenMismatch.into());
+        }
+        Ok(&message[drop_length..])
+    }
+
+    /// Read 'length' bytes of the message after the first 'drop_length' bytes
+    fn read_middle(message: &Vec<u8>, drop_length: usize, length: usize) -> Result<&[u8]> {
+        if message.len() < drop_length + length {
+            return Err(XXError::MessageLenMismatch.into());
+        }
+        Ok(&message[drop_length..(drop_length + length)])
+    }
+
+    /// Read the bytes of a key at the beginning of a message
+    fn read_key(message: &Vec<u8>) -> Result<&[u8]> {
+        Self::read_start(message, Self::key_size())
+    }
+
+    /// Size of a public key
+    fn key_size() -> usize {
+        CURVE25519_PUBLIC_LENGTH_USIZE
+    }
+
+    /// Size of an encrypted key
+    fn encrypted_key_size() -> usize {
+        Self::key_size() + AES_GCM_TAGSIZE_USIZE
+    }
+}
+
+/// The `HandshakeState` contains all the variables necessary to follow the Noise protocol
+#[derive(Debug, Clone)]
+pub(super) struct HandshakeState {
+    pub(super) s: KeyId,
+    e: KeyId,
+    k: KeyId,
+    re: PublicKey,
+    pub(super) rs: PublicKey,
+    n: u16,
+    h: [u8; SHA256_SIZE_USIZE],
+    ck: KeyId,
+    message1_payload: Vec<u8>,
+    pub(super) status: Status,
+}
+
+impl HandshakeState {
+    /// Create a new HandshakeState with:
+    ///   - a static key
+    ///   - an ephemeral key
+    ///   - an encoded identity payload (identity + signature + credentials)
+    pub(super) fn new(s: KeyId, e: KeyId, message1_payload: Vec<u8>) -> HandshakeState {
+        HandshakeState {
+            s,
+            e,
+            k: "".to_string(),
+            re: PublicKey::new(vec![], X25519),
+            rs: PublicKey::new(vec![], X25519),
+            n: 0,
+            h: [0u8; SHA256_SIZE_USIZE],
+            ck: "".to_string(),
+            message1_payload,
+            status: Initial,
+        }
+    }
+
+    /// h = SHA256(h || data)
+    pub(super) fn mix_hash(&mut self, data: &[u8]) {
+        let mut input = self.h.to_vec();
```

Comment:
Off topic for this PR, but can we make this operation more efficient? It's called pretty often

Reply:
I'm talking about allocating the memory of the right size, which we know

Reply:
Which memory allocation are you referring to, the `to_vec` part? Can you please sketch out a different implementation?

Reply:
Which memory allocation are you referring to, the `to_vec` part? Can you please sketch out a different implementation?

Reply:
```rs
let mut input = Vec::with_capacity(SHA256_SIZE_USIZE + data.len());
input.extend_from_slice(&self.h);
input.extend_from_slice(data);
self.h = Self::sha256(&input);
```

Reply:
Thanks, done.